pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,764 @@
+	if err != nil {
+		return fmt.Errorf("unable to delete %s: %v", filepath.Join(installDir, assetFilename), err)
+	}
+	return err
+}
+
+// readLinesFromFile reads lines from a file and calls the provided callback for each non-empty line.
+// It handles file opening, scanning, trimming whitespace, and error checking.
+func readLinesFromFile(filePath string, callback func(line string) error) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("could not open file: %v", err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if line == "" {
+			continue
+		}
+		if err := callback(line); err != nil {
+			return err
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("error reading file: %v", err)
+	}
+	return nil
+}
+
+func getListFromFile(path string) ([]string, error) {
+	var result []string
+	err := readLinesFromFile(path, func(line string) error {
+		if line == "/" {
+			logrus.Warnf("getListFromFile: skipping path `/`")
+			return nil
+		}
+		result = append(result, line)
+		return nil
+	})
+	if err != nil {
+		return result, fmt.Errorf("error reading asset file: %v", err)
+	}
+	return result, nil
+}
+
+// Install one or more Quadlet files
+func (ic *ContainerEngine) QuadletInstall(ctx context.Context, pathsOrURLs []string, options entities.QuadletInstallOptions) (*entities.QuadletInstallReport, error) {
+	// Is systemd available to the current user?
+	// We cannot proceed if not.
+	conn, err := systemd.ConnectToDBUS()
+	if err != nil {
+		return nil, fmt.Errorf("connecting to systemd dbus: %w", err)
+	}
+	// Is Quadlet installed? No point if not.
+	quadletPath := "/usr/lib/systemd/system-generators/podman-system-generator"
+	quadletStat, err := os.Stat(quadletPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot stat Quadlet generator, Quadlet may not be installed: %w", err)
+	}
+
+	if !quadletStat.Mode().IsRegular() || quadletStat.Mode()&0100 == 0 {
+		return nil, fmt.Errorf("no valid Quadlet binary installed to %q, unable to use Quadlet", quadletPath)
+	}
+
+	installDir := systemdquadlet.GetInstallUnitDirPath(rootless.IsRootless())
+	logrus.Debugf("Going to install Quadlet to directory %s", installDir)
+
+	_, err = os.Stat(installDir)
+	// Make the directory if it doesn't exist
+	if err != nil && errors.Is(err, fs.ErrNotExist) {
+		if err := os.MkdirAll(installDir, 0755); err != nil {
+			return nil, fmt.Errorf("unable to create Quadlet install path %s: %w", installDir, err)
+		}
+	} else if err != nil {
+		return nil, fmt.Errorf("unable to stat Quadlet install path %s: %w", installDir, err)
+	}
+
+	installReport := entities.QuadletInstallReport{
+		InstalledQuadlets: make(map[string]string),
+		QuadletErrors:     make(map[string]error),
+	}
+
+	assetFile := ""
+	paths := pathsOrURLs
+	if len(pathsOrURLs) > 0 && !strings.HasPrefix(pathsOrURLs[0], "http://") && !strings.HasPrefix(pathsOrURLs[0], "https://") {
```

Comment:
I am still missing an explanation why we special code the first arg to only allow directories. Why do we not just loop over all args and treat every arg the same.

Reply:
It's because we are identifying a install as `app` if first path is `directory` then we treat all other files in arguments as part of app otherwise we treat all arguments as isolated quadlets.